File: packages/python/slop-ai/src/slop_ai/discovery/tools.py
"""AI-facing discovery helpers for dynamic tools and core handlers."""

from __future__ import annotations

import re
from dataclasses import dataclass

from slop_ai.tools import affordances_to_tools, format_tree

from .models import DynamicToolEntry, DynamicToolResolution, DynamicToolSet, ToolResult

_SANITIZE_RE = re.compile(r"[^a-zA-Z0-9]")


def _sanitize_prefix(value: str) -> str:
    return _SANITIZE_RE.sub("_", value).strip("_")


def create_dynamic_tools(service: object) -> DynamicToolSet:
    """Build provider-prefixed dynamic affordance tools from connected providers."""
    entries: list[DynamicToolEntry] = []
    resolve_map: dict[str, DynamicToolResolution] = {}

    for provider in service.get_providers():
        tree = provider.consumer.get_tree(provider.subscription_id)
        if tree is None:
            continue

        prefix = _sanitize_prefix(provider.id)
        tool_set = affordances_to_tools(tree)
        for tool in tool_set.tools:
            resolution = tool_set.resolve(tool["function"]["name"])
            if resolution is None:
                continue
            dynamic_name = f"{prefix}__{tool['function']['name']}"
            entries.append(
                DynamicToolEntry(
                    name=dynamic_name,
                    description=f"[{provider.name}] {tool['function']['description']}",
                    input_schema=tool["function"]["parameters"],
                    provider_id=provider.id,
                    path=resolution.path,
                    action=resolution.action,
                )
            )
            resolve_map[dynamic_name] = DynamicToolResolution(
                provider_id=provider.id,
                path=resolution.path,
                action=resolution.action,
            )

    return DynamicToolSet(tools=entries, _resolve_map=resolve_map)


@dataclass(slots=True)
class ToolHandlers:
    """Core host-agnostic discovery tool handlers."""

    service: object

    async def list_apps(self) -> ToolResult:
        discovered = self.service.get_discovered()
        if not discovered:
            return ToolResult(
                content=[
                    {
                        "type": "text",
                        "text": "No applications found. Desktop and web apps that support external control will appear here automatically when they're running.",
                    }
                ]
            )

        connected = {provider.id: provider for provider in self.service.get_providers()}
        lines: list[str] = []
        for desc in discovered:
            provider = connected.get(desc.id)
            tree = (
                provider.consumer.get_tree(provider.subscription_id)
                if provider
                else None
            )
            action_count = len(affordances_to_tools(tree).tools) if tree else 0
            label = tree.properties.get("label") if tree and tree.properties else None
            if not isinstance(label, str) or not label:
                label = desc.name
            status = f"connected, {action_count} actions" if provider else "available"
            lines.append(
                f"- **{label}** (id: `{desc.id}`, {desc.transport.type}) - {status}"
            )

        return ToolResult(
            content=[
                {
                    "type": "text",
                    "text": "Applications on this computer:\n"
                    + "\n".join(lines)
                    + "\n\nUse connect_app with an app name or ID to connect and inspect it.",
                }
            ]
        )

    async def connect_app(self, app: str) -> ToolResult:
        provider = await self.service.ensure_connected(app)
        if provider is None:
            available = (
                ", ".join(
                    f"{desc.name} ({desc.id})" for desc in self.service.get_discovered()
                )
                or "none"
            )
            return ToolResult(
                content=[
                    {
                        "type": "text",
                        "text": f'App "{app}" not found. Available: {available}',
                    }
                ],
                is_error=True,
            )

        tree = provider.consumer.get_tree(provider.subscription_id)
        if tree is None:
            return ToolResult(
                content=[
                    {
                        "type": "text",
                        "text": f"{provider.name} is connected but has no state yet.",
                    }
                ]
            )

        tool_set = affordances_to_tools(tree)
        actions_text = "\n".join(
            f"  - **{(resolution.action if resolution else tool['function']['name'])}** on `{(resolution.path if resolution else '/')}`: {tool['function']['description']}"
            for tool in tool_set.tools
            for resolution in [tool_set.resolve(tool["function"]["name"])]
        )

        return ToolResult(
            content=[
                {
                    "type": "text",
                    "text": (
                        f"## {provider.name}\n"
                        f"ID: `{provider.id}`\n\n"
                        f"### Current State\n```\n{format_tree(tree)}\n```\n\n"
                        f"### Available Actions ({len(tool_set.tools)})\n{actions_text}"
                    ),
                }
            ]
        )

    async def disconnect_app(self, app: str) -> ToolResult:
        if not self.service.disconnect(app):
            return ToolResult(
                content=[
                    {
                        "type": "text",
                        "text": f'App "{app}" is not connected. Use list_apps to see available apps.',
                    }
                ],
                is_error=True,
            )

        return ToolResult(
            content=[
                {
                    "type": "text",
                    "text": f'Disconnected from "{app}". Its tools have been removed.',
                }
            ]
        )


def create_tool_handlers(service: object) -> ToolHandlers:
    """Create the core discovery tool handlers."""
    return ToolHandlers(service=service)

File: packages/typescript/integrations/discovery/src/tools.ts
import { formatTree, affordancesToTools, type LlmTool } from "@slop-ai/consumer";
import type { DiscoveryService, ConnectedProvider } from "./discovery";

export interface ToolResult {
  [key: string]: unknown;
  content: { type: "text"; text: string }[];
  isError?: boolean;
}

// ---------------------------------------------------------------------------
// Dynamic tool generation — turns affordances into per-app MCP tools
// ---------------------------------------------------------------------------

/** Sanitize a string for use as a tool name prefix (alphanumeric + underscore). */
function sanitizePrefix(s: string): string {
  return s.replace(/[^a-zA-Z0-9]/g, "_").replace(/_+/g, "_").replace(/^_|_$/g, "");
}

export interface DynamicToolEntry {
  /** Full MCP tool name, e.g. "kanban__delete" or "kanban__board__clear" */
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
  /** Provider this tool belongs to */
  providerId: string;
  /** SLOP path for invoke (null for grouped tools — path comes from args.target) */
  path: string | null;
  action: string;
  /** Valid target paths for grouped tools */
  targets?: string[];
}

export interface DynamicToolSet {
  /** All dynamic tool entries across all connected providers */
  tools: DynamicToolEntry[];
  /** Resolve a dynamic tool name to invoke coordinates, or null */
  resolve(toolName: string): {
    providerId: string;
    path: string | null;
    action: string;
    targets?: string[];
  } | null;
}

/**
 * Build dynamic tool definitions from all connected providers' affordances.
 *
 * Each tool name is prefixed with the app's sanitized ID to avoid cross-app collisions:
 *   `{appPrefix}__{action}` (grouped) or `{appPrefix}__{nodeId}__{action}` (singleton)
 *
 * Grouped tools share the same action + param schema across multiple nodes.
 * They include a `target` parameter so the caller specifies which node to act on.
 *
 * Returns a DynamicToolSet with all tools and a resolve function.
 */
export function createDynamicTools(discovery: DiscoveryService): DynamicToolSet {
  const entries: DynamicToolEntry[] = [];
  const resolveMap = new Map<string, { providerId: string; path: string | null; action: string; targets?: string[] }>();

  for (const provider of discovery.getProviders()) {
    const tree = provider.consumer.getTree(provider.subscriptionId);
    if (!tree) continue;

    const appPrefix = sanitizePrefix(provider.id);
    const toolSet = affordancesToTools(tree);

    for (const tool of toolSet.tools) {
      const resolved = toolSet.resolve(tool.function.name);
      if (!resolved) continue;

      const dynamicName = `${appPrefix}__${tool.function.name}`;

      entries.push({
        name: dynamicName,
        description: `[${provider.name}] ${tool.function.description}`,
        inputSchema: tool.function.parameters,
        providerId: provider.id,
        path: resolved.path,
        action: resolved.action,
        targets: resolved.targets,
      });

      resolveMap.set(dynamicName, {
        providerId: provider.id,
        path: resolved.path,
        action: resolved.action,
        targets: resolved.targets,
      });
    }
  }

  return {
    tools: entries,
    resolve(toolName: string) {
      return resolveMap.get(toolName) ?? null;
    },
  };
}

export function createToolHandlers(discovery: DiscoveryService) {
  async function listApps(): Promise<ToolResult> {
    const discovered = discovery.getDiscovered();
    if (discovered.length === 0) {
      return {
        content: [
          {
            type: "text",
            text: "No applications found. Desktop and web apps that support external control will appear here automatically when they're running.",
          },
        ],
      };
    }

    const connected = discovery.getProviders();
    const connectedIds = new Set(connected.map((p) => p.id));

    const lines = discovered.map((desc) => {
      const isConnected = connectedIds.has(desc.id);
      const provider = isConnected
        ? connected.find((p) => p.id === desc.id)
        : null;
      const tree = provider?.consumer.getTree(provider.subscriptionId);
      const actionCount = tree ? affordancesToTools(tree).tools.length : 0;
      const label = (tree?.properties?.label as string) ?? desc.name;
      const status = isConnected
        ? `connected, ${actionCount} actions`
        : "available";
      return `- **${label}** (id: \`${desc.id}\`, ${desc.transport.type}) — ${status}`;
    });

    return {
      content: [
        {
          type: "text",
          text: `Applications on this computer:\n${lines.join("\n")}\n\nUse connect_app with an app name or ID to connect and inspect it.`,
        },
      ],
    };
  }

  async function connectApp(args: { app: string }): Promise<ToolResult> {
    const { app } = args;

    // --- Show specific app (lazy connect) ---
    const p = await discovery.ensureConnected(app);
    if (!p) {
      const discovered = discovery.getDiscovered();
      const available = discovered
        .map((d) => `${d.name} (${d.id})`)
        .join(", ");
      return {
        content: [
          {
            type: "text",
            text: `App "${app}" not found. Available: ${available || "none"}`,
          },
        ],
      };
    }

    const tree = p.consumer.getTree(p.subscriptionId);
    if (!tree) {
      return {
        content: [
          {
            type: "text",
            text: `${p.name} is connected but has no state yet.`,
          },
        ],
      };
    }

    const toolSet = affordancesToTools(tree);
    const actionsText = toolSet.tools
      .map((t) => {
        const resolved = toolSet.resolve(t.function.name);
        const action = resolved?.action ?? t.function.name;
        const pathInfo = resolved?.path
          ? `on \`${resolved.path}\``
          : `${resolved?.targets?.length ?? 0} targets`;
        return `  - **${action}** ${pathInfo}: ${t.function.description}`;
      })
      .join("\n");

    return {
      content: [
        {
          type: "text",
          text:
            `## ${p.name}\nID: \`${p.id}\`\n\n` +
            `### Current State\n\`\`\`\n${formatTree(tree)}\n\`\`\`\n\n` +
            `### Available Actions (${toolSet.tools.length})\n${actionsText}`,
        },
      ],
    };
  }

  async function disconnectApp(args: { app: string }): Promise<ToolResult> {
    const { app } = args;
    const found = discovery.disconnect(app);
    if (!found) {
      return {
        content: [
          {
            type: "text",
            text: `App "${app}" is not connected. Use list_apps to see available apps.`,
          },
        ],
      };
    }
    return {
      content: [
        {
          type: "text",
          text: `Disconnected from "${app}". Its tools have been removed.`,
        },
      ],
    };
  }

  return { listApps, connectApp, disconnectApp };
}
